User Profile: Initial beta overview documentations (#83766)
elastic/elasticsearch: x-pack/docs/en/security/authentication/security-domain.asciidoc

[role="xpack"]
[[security-domain]]
=== Security domains

beta::[]

Security domains are a method of grouping <<realms,realms>> to enhance resource 
sharing and check ownership for the same user across different realms.

Users can create various resources in {es} that they own, access, and manage.
For example, using the <<async-search,async search API>> to create a search 
request, <<security-api-create-api-key,creating an API key>>, or activating a 
<<user-profile,user profile>>.

When a user attempts to access a resource, {es} compares a user's `username`
and the realm they're authenticating from to determine who they are. In some
cases, the same user can authenticate with different realms. For example,
an <<ldap-realm,LDAP realm>> and a <<saml-realm,SAML realm>> can be backed by
the same directory service. Similarly, a SAML realm might
<<configuring-authorization-delegation,delegate authorization>> to a

> <<native-realm,native realm>>.

Reviewer: @ywangd Is this supposed to refer to the fact that users authn to a SAML realm which delegates authorization to an LDAP realm, are viewed as different users if they would directly authn to the LDAP realm, because I don't believe this is the case.

If {es} detects that a `username` from two different realms is attempting to

> access a resource, {es} assumes that these users are unique and doesn't allow

Reviewer: ```suggestion
access a resource, {es} assumes that these users are distinct and doesn't allow
````

> resources to be shared between them. With security domains, you can associate a 

Author: At high level, associating a single user profile to multiple logins of the same user is also resource sharing. So I don't consider it being a separate functions, but rather one concrete application of the function, other applications are async search result, API keys etc.

single <<user-profile,user profile>> with a `username` to enable resource sharing.
This means that user preferences, notifications, and other user data can be
shared across domains. The same user can view results from an asynchronous

> search request or a scrolling search, regardless of which domain they're 

Reviewer: ```suggestion
search request or a scrolling search, regardless of which realm under the same domain they're 
```

[role="xpack"]
[[security-domain]]
=== Security domains

beta::[]

> Security domains are a method of grouping <<realms,realms>> to enable resource
> sharing across different realms for the same user.

Reviewer 2: This wording "different realms for the same user" is a little awkward, but it's hard to come up with something better.

I think perhaps the problem is trying to describe it as "resource sharing" which is a little too much _implementation_. If we try to approach it from a conceptual point of view we can word things differently.
 
```suggestion
Security domains are a method of grouping <<realms,realms>> so that the {stack}
can recognize when a single user authenticates to two different realms.
```
I still don't love it, but I think it's easier to introduce the concept that way first.

Author: I modified your version a little bit to say
> Security domains are a method of grouping different <<realms,realms>>
so that the {stack} can recognize when a single user authenticates with
these realms.

> ==== How security domain works

Reviewer 2: ```suggestion
==== How security domains work
```

Alternatively "How __*a*__ security domain works", but that doesn't read as well.

> Users can create various resources in {es} that they own, access, and manage.
> For example, using the <<async-search,async search API>> to create a search
> request, <<security-api-create-api-key,creating an API key>>, or activating a
> <<user-profile,user profile>>.

Reviewer 2: Strictly, users don't access or manage their own profiles (nor can they "create" them)

Perhaps we should say
```suggestion
Some types of resources in {es} are owned by a single user.
For example <<async-search,async search contexts>>,
<<security-api-create-api-key,API keys>>,
and <<user-profile,user profiles>> each have an owning user.
```

Author: Good catch!

When a user creates a resource, {es} captures the user's `username` and realm
information as part of the resource's metadata.
When the user attempts to access the resource at a later time, {es} compares

> captured `username` and realm information against those from the accessing

Reviewer 2: ```suggestion
the captured `username` and realm information against those from the accessing
```

> user. The access is allowed only when both of them match.

Reviewer 2: ```suggestion
user. The access is not allowed unless both the realm and `username` match.
```

That is, if {es} detects that a `username` from two different realms is
attempting to access a resource, {es} assumes that these users are distinct and
doesn't allow resources to be shared between them.

> However, there are cases where the same end-user can authenticate with

Reviewer 2: ```suggestion
However, there are cases where the same individual is able to authenticate with
```

different realms and need to share the same set of resources across the realms.
For example, an <<ldap-realm,LDAP realm>> and a <<saml-realm,SAML realm>> can
be backed by the same directory service. Additionally,
<<configuring-authorization-delegation,authorization delegation>> allows one
realm to delegate authorization to another realm. If both realms authenticate
end-users with the same `username`, it is reasonable to treat these users to
be the same user.

With security domains, you can configure multiple realms to be under the same
domain, users can authenticate with any of the realms and have access to the

> same set of resources regardless of which realm under the same domain they're
> authenticating from. For example, a single <<user-profile,user profile>> is

Reviewer 2: I don't feel strongly about this one, but the sentence was long and I suspect it could hard to follow for people with a lower proficiency in English.

```suggestion
same set of resources regardless of which realm authenticating from, 
(provided those realms are under the same domain).
For example, a single <<user-profile,user profile>> is
```

> authenticating from. For example, a single <<user-profile,user profile>> is
> associated to the user which enables user preferences, notifications, and other
> user data to be shared across domains. The same user can view results from an

Reviewer 2: I found the "user which enables" to be hard to follow.
We could change it to "user, enabling ..."

Or something more like this:
```suggestion
authenticating from.
For example, two users who are linked by a single domain will
share a <<user-profile,user profile>>, and can share user
preferences, notifications, and other user data. 
The same user can view results from an
```

Author: I updated to `user, enabling ...` because your other version uses the term `link` and I am concerned this might be another cogonitive load for readers.

asynchronous search request or a scrolling search. If the user has the
necessary privileges, they can also view and manage API keys across realms.

==== Configure a security domain

You specify the realms for a domain in `elasticsearch.yml`. The realms must

> be defined in `elasticsearch.yml`, it doesn't have to be enabled to be listed
> under a domain.

Reviewer 2: ```suggestion
be defined in `elasticsearch.yml` (but do not need to be enabled).
```

A realm doesn't need to belong to a domain. If it does, the realm can only
belong to a single domain. Specifying a realm in multiple domains generates an
error when {es} starts.

NOTE: The <<file-realm,file realm>> and <<native-realm,native realm>> are
automatically enabled as `default_file`, and `default_native`, respectively,
without the need for explicit configuration. They can
be listed under domains even when they are not explicitly defined in
`elasticsearch.yml`.

. Add a security domain configuration to `elasticsearch.yml` under the
`xpack.security.authc.domains` namespace:
+
[source, yaml]
----
xpack:
  security:
    authc:
      domains:
        my_domain:
          realms: [ 'default_native', 'saml1' ] <1>
----
<1> This configuration defines a security domain called `my_domain`, which
contains two realms named `default_native` and `saml1`.

. Restart {es}.

[NOTE]
====
Though it's possible to configure different <<roles,roles>> for the same user when
authenticating with different realms, it is _not_ recommended.
Using <<configuring-authorization-delegation,authorization delegation>>
is a good way to ensure that a user is assigned the same roles from multiple realms.

> {es} always enforces the user privileges for the authentication it has at the moment.

Reviewer 2: ```suggestion
{es} always enforces the privileges associated with the currently authenticated user.
```

For example <<async-search,async search contexts>>,
<<security-api-create-api-key,API keys>>,
and <<user-profile,user profiles>> each have an owning user.

When a user creates a resource, {es} captures the user's `username` and realm
information as part of the resource's metadata.
When the user attempts to access the resource at a later time, {es} compares
the captured `username` and realm information against those from the accessing
user. The access is not allowed unless both the realm and `username` match.
That is, if {es} detects that a `username` from two different realms is
attempting to access a resource, {es} assumes that these users are distinct and
doesn't allow resources to be shared between them.

However, there are cases where the same user can authenticate with
different realms and need to share the same set of resources across the realms.
For example, an <<ldap-realm,LDAP realm>> and a <<saml-realm,SAML realm>> can
be backed by the same directory service. Additionally,
<<configuring-authorization-delegation,authorization delegation>> allows one
realm to delegate authorization to another realm. If both realms authenticate

> users with the same `username`, it is reasonable to treat these users to
> be the same user.

Reviewer: ```suggestion
users with the same `username`, it is reasonable to treat these users as the
same end-user identity from a resource ownership point of view.
```

> With security domains, you can configure multiple realms to be under the same
> domain, users can authenticate with any of the realms and have access to the

Reviewer: ```suggestion
With security domains, you group multiple realms to be part of a domain. Users can authenticate with any of the domain realms and have access to the
```

Author: I prefer keep saying just `any of the realms` instead of `any of the domain realms` since the meaning is clear enough given the context. Even though it is clear to us what `domain realms` means, it could still be confusing to other readers. So I'd avoid it unless we need to formally introduce the term.
I took other changes. Thanks!

> same set of resources regardless of which realm authenticating from,

Reviewer: ```suggestion
same set of resources regardless of which realm they authenticated with,
```

(provided those realms are under the same domain).
For example, a single <<user-profile,user profile>> is

> associated to the user, enabling user preferences, notifications, and other
> user data to be shared across domains. The same user can view results from an

Reviewer: ```suggestion
associated to the end-user identity, enabling preferences, notifications, and other
user data to be shared across domain realms. The single user can view results from an
```

asynchronous search request or a scrolling search. If the user has the
necessary privileges, they can also view and manage API keys across realms.

==== Configure a security domain

> You specify the realms for a domain in `elasticsearch.yml`. The realms must

Reviewer: ```suggestion
You associate realms under a domain in `elasticsearch.yml`. All the domain realms must
```

be defined in `elasticsearch.yml` (but do not need to be enabled).

> Specifying either an undefined realm, synthetic realm, or the reserved realm
> under a domain will prevent the node from starting.
> A realm doesn't need to belong to a domain. If it does, the realm can only
> belong to a single domain. Specifying a realm in multiple domains generates an
> error when {es} starts.

Reviewer: ```suggestion
Realms can be associated to at most one domain. It is an error ({es} fails to start) to assign an undefined realm to a domain or to assign a single realm to multiple domains.
```

Author: I reordered this section and moved it under the configuration example which has a second step says "Restart ES". It feels more naturally to talk about what could cause failure to start right after the instruction of restart. It also reads better to give example first before talking about error conditions.

For example, an <<ldap-realm,LDAP realm>> and a <<saml-realm,SAML realm>> can
be backed by the same directory service. Additionally,
<<configuring-authorization-delegation,authorization delegation>> allows one
realm to delegate authorization to another realm. If both realms authenticate
users with the same `username`, it is reasonable to treat these users as the
same end-user from a resource ownership point of view.

With security domains, you group multiple realms to be under the same
domain, end-users can authenticate with any of the realms and have access to the
same set of resources regardless of which realm they authenticated with,
(provided those realms are under the same domain).
For example, a single <<user-profile,user profile>> is
associated to the end-user, enabling preferences, notifications, and other
user data to be shared across realms. The user can view results from an
asynchronous search request or a scrolling search across realms. If the user has the
necessary privileges, they can also view and manage API keys across realms.

==== Configure a security domain

You configure realms to be under a domain in `elasticsearch.yml`. The realms must

> be defined in `elasticsearch.yml` (but do not need to be enabled).

Author: When a realm is defined in the yaml file, it is automatically enabled, but you can opt-in to disable it with `enabled: false`, e.g.:
```
ldap.ldap1:
  enabled: false
  ...
```

NOTE: The <<file-realm,file realm>> and <<native-realm,native realm>> are
automatically enabled as `default_file`, and `default_native`, respectively,
without the need for explicit configuration. They can
be listed under domains even when they are not explicitly defined in
`elasticsearch.yml`.

. Add a security domain configuration to `elasticsearch.yml` under the
`xpack.security.authc.domains` namespace:
+
[source, yaml]
----
xpack:
  security:
    authc:
      domains:
        my_domain:
          realms: [ 'default_native', 'saml1' ] <1>
----
<1> This configuration defines a security domain called `my_domain`, which
contains two realms named `default_native` and `saml1`.

. Restart {es}.

{es} can fail to start on invalid domain configuration including the followings:
* A single realm is configured to be under multiple domains
* Any undefined realm, or synthetic realm, or the reserved realm is configured to
  be under a domain

> [NOTE]
> ====
> Though it's possible to configure different <<roles,roles>> for the same user when
> authenticating with different realms, it is _not_ recommended.
> Using <<configuring-authorization-delegation,authorization delegation>>
> is a good way to ensure that a user is assigned the same roles from multiple realms.
> 
> {es} always enforces the privileges associated with the currently authenticated user.
> Security domains don't bypass <<authorization,user authorization>> when
> resource sharing requires them. For example, a user requires the
> `manage_own_api_key` privilege to manage their own API keys. If that user does
> not have this privilege when authenticating with one realm, they won't be able
> to manage API keys while authenticating with this realm.
> This behaviour might be unexpected since the same user may or may not be able
> to perform the same operation depending on the realm it authenticates with.
> But it is a safe behaviour that ensures actual actions always comply to
> authorization requirements.
> ====

Reviewer 3: ```suggestion
.Managing privileges across realms
***
Though it's possible to configure different <<roles,roles>> for the same user when
authenticating with different realms, it is _not_ recommended.
Using <<configuring-authorization-delegation,authorization delegation>>
is a good way to ensure that a user is assigned the same roles from multiple realms.

{es} always enforces the privileges associated with the currently authenticated user.
Security domains don't bypass <<authorization,user authorization>> when
resource sharing requires them. For example, a user requires the
`manage_own_api_key` privilege to manage their own API keys. If that user doesn't
have this privilege when authenticating with one realm, they won't be able
to manage API keys while authenticating with this realm.
This behavior might be unexpected since a single user may or may not be able
to perform the same operation depending on the realm it authenticates with.
However, it's a safe behavior that ensures actual actions always comply to
authorization requirements.
****
```
